airflow/operators/singularity_operator.py

```diff
@@ -0,0 +1,179 @@
+# under the License.
+
+import json
+
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+from airflow.utils.file import TemporaryDirectory
+from spython.main import Client
+import shutil
+import ast
+import os
+
+
+class SingularityOperator(BaseOperator):
+    """
+    Execute a command inside a Singularity container
+
+    Singularity has more seamless connection to the host than Docker, so
+    no special binds are needed to ensure binding content in the user $HOME
+    and temporary directories. If the user needs custom binds, this can
+    be done with --volumes
+
+    :param image: Singularity image or URI from which to create the container.
+    :type image: str
+    :param auto_remove: Delete the container when the process exits
+                        The default is False.
+    :type auto_remove: bool
+    :param command: Command to be run in the container. (templated)
+    :type command: str or list
+    :param start_command: start command to pass to the container instance
+    :type start_command: string or list
+    :param environment: Environment variables to set in the container. (templated)
+    :type environment: dict
+    :param force_pull: Pull the image on every run. Default is False.
+    :type force_pull: bool
+    :param volumes: List of volumes to mount into the container, e.g.
+        ``['/host/path:/container/path', '/host/path2:/container/path2']``.
+    :param options: other flags (list) to provide to the instance start
+    :type options: list
+    :param working_dir: Working directory to
+        set on the container (equivalent to the -w switch the docker client)
+    :type working_dir: str
+    """
+    template_fields = ('command', 'environment',)
+    template_ext = ('.sh', '.bash',)
+
+    @apply_defaults
+    def __init__(
+            self,
+            image,
+            api_version=None,
+            command=None,
+            start_command=None,
+            environment=None,
+            pull_folder=None,
+            force_pull=False,
+            volumes=None,
+            options = None,
+            working_dir=None,
+            auto_remove=False,
+            *args,
+            **kwargs):
+
+        super(SingularityOperator, self).__init__(*args, **kwargs)
+        self.api_version = api_version
+        self.auto_remove = auto_remove
+        self.command = command
+        self.start_command = start_command
+        self.environment = environment or {}
+        self.force_pull = force_pull
+        self.image = image
+        self.instance = None
+        self.options = options or []
+        self.volumes = volumes or []
+        self.working_dir = working_dir
+        self.cli = None
+        self.container = None
+
+    def execute(self, context):
+        self.log.info('Preparing Singularity container %s', self.image)
+        self.cli = Client
```

Comment:
The operator should not call the external library directly. It is recommended to introduce an intermediate layer - hook.

Reply:
I'm not sure I understand - in my case, the spython.main.Client is a python wrapper already to interact with Singularity. What purpose does hook serve?

Reply:
I looked deeper and this library does not take parameters in constructor, so there is no justification. Sorry.

Reference: 
https://singularityhub.github.io/singularity-cli/api/modules/spython/main.html

Reply:
I'm not understanding how we can actually test the Singularity operator in isolation of having Singularity. What it sounds like you are suggesting is to test the Client that runs the operator, but isn't this redundant if the client already has all tests?

Where I'm at - the example workflow with Singularity works, but I haven't a clue how to write tests for it with this mock setup.

Reply:
You should only test your operator. The tests should only cover the classes that you have written. Client should be mocked.

Reply:
I don't understand the mock - I've been trying to write it following the other examples but I don't follow exactly how it's working, or what I should do based on the error messages. For example, here I'm trying to test a "mock" of execute:

```
class SingularityOperatorTestCase(unittest.TestCase):
    @mock.patch('airflow.utils.file.mkdtemp')
    @mock.patch('airflow.contrib.operators.singularity_operator.Client')
    def test_execute(self, client_class_mock, mkdtemp_mock):
        host_config = mock.Mock()
        mkdtemp_mock.return_value = '/mkdtemp'

        client_mock = mock.Mock(spec=Client)

        #client_mock.instance.start.return_value = 0
        client_mock.instance.stop.return_value = 0

        operator = SingularityOperator(command='env', 
                                       environment={'UNIT': 'TEST'},
                                       image='docker://busybox:1.30.1',
                                       task_id='unittest')
        operator.execute(None)

        #client_mock.instance.assert_called_with(image='docker://busybox:1.30.1',
        #                                        start=False,
        #                                        args=None,
        #                                        options=['--bind', '/host/path:/container/path', '--workdir', '/container/path'])

        #client_mock.execute.assert_called_with(client_mock.instance,
        #                                       command='env',
        #                                       return_result=True)
```
And running the test, I'm not sure what is wrong:

```python
$ python -m unittest test_singularity_operator
E
======================================================================
ERROR: test_execute (test_singularity_operator.SingularityOperatorTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vanessa/anaconda3/lib/python3.6/unittest/mock.py", line 1179, in patched
    return func(*args, **keywargs)
  File "/home/vanessa/Documents/Dropbox/Code/labs/cherry/airflow/tests/contrib/operators/test_singularity_operator.py", line 56, in test_execute
    operator.execute(None)
  File "/home/vanessa/anaconda3/lib/python3.6/site-packages/airflow/contrib/operators/singularity_operator.py", line 160, in execute
    raise AirflowException('Singularity failed: %s' % message)
airflow.exceptions.AirflowException: Singularity failed: <MagicMock name='Client.execute().__getitem__()' id='139677815615272'>

----------------------------------------------------------------------
Ran 1 test in 0.004s

FAILED (errors=1)
```
This is probably my misunderstanding of the mock library / mocking in general - any help that you can provide (a suggestion to what I'm doing above?) would be very helpful.

Reply:
Calling the method from the library returns the result, which is then used in the code. If you do not specify it, the Mock object is returned.
You must add a piece of code:
```python
client_mock.execute.return_value = dict(return_code=0, message="message")
```

This code snippet is not used. You have a mock object in arguments of method.
``
client_mock = mock.Mock(spec=Client)
``